src/archive/src/main.rs

```diff
@@ -0,0 +1,394 @@
+        buf.extend(&self.log_index.to_le_bytes());
+        assert_eq!(buf.len(), USER_INDEX_KEY_LENGTH);
+        Cow::Owned(buf)
+    }
+
+    fn from_bytes(bytes: Vec<u8>) -> Self {
+        assert_eq!(bytes.len(), USER_INDEX_KEY_LENGTH);
+        UserIndexKey {
+            user_number: u64::from_le_bytes(
+                TryFrom::try_from(&bytes[0..8]).expect("failed to read user_number"),
+            ),
+            timestamp: u64::from_le_bytes(
+                TryFrom::try_from(&bytes[8..16]).expect("failed to read timestamp"),
+            ),
+            log_index: u64::from_le_bytes(
+                TryFrom::try_from(&bytes[16..]).expect("failed to read log_index"),
+            ),
+        }
+    }
+}
+
+#[update]
+fn write_entry(user_number: UserNumber, timestamp: Timestamp, entry: ByteBuf) {
+    with_config(|config| {
+        if config.ii_canister != caller() {
+            trap(&format!(
+                "Only {} is allowed to write entries.",
+                config.ii_canister
+            ))
+        }
+    });
+    let idx = with_log(|log| {
+        log.append(entry.as_ref())
+            .expect("failed to append log entry")
+    });
+
+    with_user_index(|index| {
+        let key = UserIndexKey {
+            user_number,
+            timestamp,
+            log_index: idx as u64,
+        };
+
+        // the only way this expect can trigger is when the key size is wrong.
+        // On other failures (e.g. no more stable memory available) the underlying StableBTreeMap will panic directly.
+        index.insert(key, ()).expect("bug: key size mismatch");
+    })
+}
+
+#[query]
+fn get_logs(index: Option<u64>, limit: Option<u16>) -> Logs {
+    let num_entries = sanitize_limit(limit);
+
+    with_log(|log| {
+        let length = log.len();
+        let start_idx = match index {
+            None => length.saturating_sub(num_entries),
+            Some(idx) => idx as usize,
+        };
+
+        let next_idx = if start_idx + num_entries < length {
+            Some((start_idx + num_entries) as u64)
+        } else {
+            None
+        };
+
+        let mut entries = Vec::with_capacity(log.len().min(num_entries));
+        for idx in start_idx..start_idx + num_entries {
+            let entry = match log.get(idx) {
+                None => break,
+                Some(entry) => entry,
+            };
+            entries.push(Some(
+                candid::decode_one(&entry).expect("failed to decode log entry"),
+            ))
+        }
+        Logs { entries, next_idx }
+    })
+}
+
+#[query]
+fn get_user_logs(user_number: u64, cursor: Option<Cursor>, limit: Option<u16>) -> UserLogs {
+    let num_entries = sanitize_limit(limit);
+
+    with_user_index(|index| {
+        let iterator = match cursor {
+            None => index.range(user_number.to_le_bytes().to_vec(), None),
+            Some(Cursor::NextToken { next_token }) => index.range(next_token.into_vec(), None),
```

Comment:
do these implicitly rely on the `Storable` implementation?

Reply:
Yes, they are tied to the `AnchorIndexKey`. The `next_token` is directly the byte representation obtained by the `Storable` implementation.